### PR TITLE
Experimental Transfer table and Management API

### PR DESCRIPTION
**Pull request by ysv**

### CODE REVIEW

File: app/api/v2/management/entities/transfer.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                   }
+          end
+        end
+      end
+    end
+  end
+end
+
+
+# == Schema Information
```
Comment (ysv) [inline app/api/v2/management/entities/transfer.rb]: Remove this

File: app/api/v2/management/entities/transfer.rb [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                   desc: 'Transfer Kind.'
+                 }
+
+          expose :desc,
+                 documentation: {
+                   type: String,
+                   desc: 'Transfer Description'
+                 }
+
+          # Expose assets, expenses, liabilities, revenues.
```
Comment (ysv) [inline app/api/v2/management/entities/transfer.rb]: Expose assets, expenses, liabilities, revenues if they are present

File: app/api/v2/management/entities/transfer.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+          ::Operation::TYPES.map(&:pluralize).each do |op_t|
+            expose op_t,
+                   using: Operation,
+                   if: ->(transfer) { transfer.send(op_t).present? },
```
Comment (ysv) [inline app/api/v2/management/entities/transfer.rb]: public_send

File: app/api/v2/management/entities/transfer.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        class Transfer < Base
+          expose :key,
+                 documentation: {
+                   type: Integer,
+                   desc: 'Unique Transfer Key.'
+                 }
+
+          expose :kind,
+                 documentation: {
+                   type: Integer,
```
Comment (ysv) [inline app/api/v2/management/entities/transfer.rb]: String

File: app/api/v2/management/helpers.rb [hunk 5]
```diff
@@ -5,7 +5,105 @@ module API
...
+                                 member_id: member.id,
+                                 reference: attrs[:reference])
+
+              credit_legacy_balance!(amount: amount,
+                                     member: member,
+                                     currency: currency,
+                                     kind: op.account.kind)
+              op
+            end
+          elsif attrs[:debit].present?
```
Comment (ysv) [inline app/api/v2/management/helpers.rb]: Check indent

File: app/api/v2/management/helpers.rb [hunk 6]
```diff
@@ -5,7 +5,105 @@ module API
...
+            member.ac(currency).plus_funds(amount)
+          elsif kind.to_s == 'locked'
+            member.ac(currency).plus_funds(amount)
+            member.ac(currency).lock_funds(amount)
+          end
+        end
+
+        # @deprecated
+        def debit_legacy_balance!(amount:, member:, currency:, kind:)
+
```
Comment (ysv) [inline app/api/v2/management/helpers.rb]: blank line

File: app/api/v2/management/helpers.rb [hunk 7]
```diff
@@ -5,7 +5,105 @@ module API
...
+                                    currency: currency,
+                                    kind: op.account.kind)
+              op
+            end
+          end
+        end
+
+        # @deprecated
+        def credit_legacy_balance!(amount:, member:, currency:, kind:)
+          kind ||= ::Operations::Chart.find_account_by(code: code).fetch(:kind)
```
Comment (ysv) [inline app/api/v2/management/helpers.rb]: remove this

File: app/api/v2/management/operations.rb [hunk 8]
```diff
@@ -139,9 +135,15 @@ class Operations < Grape::API
...
                      values: -> { ::Currency.codes(bothcase: true) },
                      desc: 'The currency code.'
-            optional :uid,
-                     type: String,
-                     desc: 'The user ID for operation owner.'
+            requires :code,
+                     type: Integer,
+                     values: -> { ::Operations::Chart.codes(type: op_type) },
+                     desc: 'Operation account code'
+            given code: ->(code) { ::Operations::Chart.find_account_by(code: code).fetch(:scope) == 'member' } do
```
Comment (ysv) [inline app/api/v2/management/operations.rb]: fetch(:scope).to_s

File: app/api/v2/management/transfers.rb [hunk 9]
```diff
@@ -0,0 +1,94 @@
+# encoding: UTF-8
+# frozen_string_literal: true
+
+module API
+  module V2
+    module Management
+      class Transfers < Grape::API
+
```
Comment (ysv) [inline app/api/v2/management/transfers.rb]: blank line

File: app/api/v2/management/transfers.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+              end
+            end
+          end
+        end
+        post '/transfers/new' do
+          declared_params = declared(params)
+
+          Transfer.transaction do
+            transfer = Transfer.create!(declared_params.slice(:key, :kind, :desc))
+            declared_params[:operations].map do |pair|
```
Comment (ysv) [inline app/api/v2/management/transfers.rb]: rename pair

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -81,7 +81,7 @@ def set_gon
         month: I18n.t('chart.month'),
         year: I18n.t('chart.year')
       },
-      chart: {
```
Comment (ysv) [inline app/controllers/application_controller.rb]: reset this change

File: app/models/operations/chart.rb [hunk 12]
```diff
@@ -1,102 +1,36 @@
 # encoding: UTF-8
 # frozen_string_literal: true
 
+# TODO: Move it to ActiveRecord model.
```
Comment (ysv) [inline app/models/operations/chart.rb]: improve comment

File: spec/factories/transfer.rb [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      with_revenues
+    end
+
+    factory :transfer_with_operations, traits: %i[with_operations]
+  end
+
+end
+
+
+# == Schema Information
```
Comment (ysv) [inline spec/factories/transfer.rb]: Remove

**Comment by ianeinser** [comment]
This wont work for fiat transfer. Reasons:
1. Fiat transfer should involve real cash transfer due to some regulations to monitor terrorism and money laundry activities
2. Some users may deposit from different bank accounts of even different banks, so how Peatio can identified that and provide regular updates to authorities of transfer traffic because by allowing internal transfer, Peatio can serve like a Bank due to its centralized characteristic

**Comment by ysv** [comment]
Moved to #2023

**Comment by ysv** [comment]
@ianeinser let's discuss in #2023

**Pull request closed without merging**